rollouts/controllers/rollout_controller.go

```diff
@@ -749,13 +794,57 @@ func (r *RolloutReconciler) mapClusterUpdateToRequest(cluster client.Object) []r
 	return requests
 }
 
+func getTargetsForWave(allTargets *Targets,
+	rollout *gitopsv1alpha1.Rollout,
+	waveClusters []gkeclusterapis.ContainerCluster, waveName string) *Targets {
+	targets := &Targets{}
+
+	clusterNames := make(map[string]struct{})
+
+	for _, cluster := range waveClusters {
+		clusterNames[cluster.Name] = struct{}{}
+	}
+
+	for _, toCreate := range allTargets.ToBeCreated {
+		if _, inWave := clusterNames[toCreate.cluster.Name]; inWave {
+			targets.ToBeCreated = append(targets.ToBeCreated, toCreate)
+		}
+	}
+
+	for _, rrs := range allTargets.ToBeUpdated {
+		if _, inWave := clusterNames[rrs.Spec.ClusterRef.Name]; inWave {
+			targets.ToBeUpdated = append(targets.ToBeUpdated, rrs)
+		}
+	}
+
+	for _, rrs := range allTargets.Unchanged {
+		if _, inWave := clusterNames[rrs.Spec.ClusterRef.Name]; inWave {
+			targets.Unchanged = append(targets.Unchanged, rrs)
+		}
+	}
+
+	if len(allTargets.ToBeDeleted) > 0 && len(rollout.Status.WaveStatuses) > 0 {
```

Comment:
I couldn't understand the second condition ? (Is it safe to assume that the wave will exist in the status ? or should we attach waveName to rrs either as label or annotation ?)

Reply:
Take a look at the latest code. I was able to eliminate this condition here with new logic that is much easier to follow.